src/vos/vos_dtx_cos.c

```diff
@@ -360,42 +355,78 @@ vos_dtx_list_cos(daos_handle_t coh, daos_unit_oid_t *oid, uint64_t dkey_hash,
 
 int
 vos_dtx_add_cos(daos_handle_t coh, daos_unit_oid_t *oid, struct dtx_id *dti,
-		uint64_t dkey_hash, daos_epoch_t epoch, bool punch, bool check)
+		uint64_t dkey_hash, daos_epoch_t epoch, uint64_t gen,
+		bool punch)
 {
+	struct daos_lru_cache		*occ = vos_obj_cache_current();
+	struct vos_object		*obj = NULL;
 	struct vos_container		*cont;
 	struct dtx_cos_key		 key;
 	struct dtx_cos_rec_bundle	 rbund;
+	daos_epoch_range_t		 epr = {0, epoch};
 	d_iov_t				 kiov;
 	d_iov_t				 riov;
 	int				 rc;
 
 	cont = vos_hdl2cont(coh);
 	D_ASSERT(cont != NULL);
 
-	if (check) {
-		struct daos_lru_cache	*occ = vos_obj_cache_current();
-		struct vos_object	*obj = NULL;
-		daos_epoch_range_t	 epr = {0, epoch};
+	/* If the DTX is started befoe DTX resync operation (for rebuild),
+	 * then it is possbile that the DTX resync ULT may have aborted
+	 * or committed the DTX during current ULT waiting for the reply.
+	 * let's check DTX status locally before marking as 'committable'.
+	 */
+	if (gen == 0)
+		goto add;
+
+	if (gen < cont->vc_dtx_resync_gen) {
+		rc = vos_dtx_check(coh, dti);
+		switch (rc) {
+		case DTX_ST_PREPARED:
+			rc = vos_dtx_lookup_cos(coh, oid, dti,
+						dkey_hash, punch);
+			/* The resync ULT has already added it into the
+			 * CoS cache, current ULT needs to do nothing.
+			 */
+			if (rc == 0)
```

Comment:
why do we bother to return ALREADY? We could just return zero and simply caller right?

Reply:
OK